tests/lib/rules/no-restricted-imports.js

```diff
@@ -35,6 +35,134 @@ ruleTester.run("no-restricted-imports", rule, {
         {
             code: "import withGitignores from \"foo/bar\";",
             options: [{ patterns: ["foo/*", "!foo/bar"] }]
+        },
+        {
+            code: "import AllowedObject from \"foo\";",
+            options: [{
+                paths: [{
+                    name: "foo",
+                    importNames: ["DisallowedObject"]
+                }]
+            }]
+        },
+        {
+            code: "import DisallowedObject from \"foo\";",
+            options: [{
+                paths: [{
+                    name: "foo",
+                    importNames: ["DisallowedObject"]
```

Comment:
I’m confused; is the intent here to restrict the *name* the user has chosen, or to restrict the default import under any name?

Reply:
The intent is to restrict the usage of an particular export by specifying the name of that export. Using a different local name for the export should still cause an error to get reported. Updating the external module to export the same thing under a different name would prevent an error from getting reported, although this would not be recommended because it would effectively defeat the point of the rule.

Reply:
Does that mean this rule parses the imported file, when checking the importing file?

Reply:
No, it only parses the importing file. I meant that if someone has code like this:

```js
// external-module.mjs

export function bad() {
  // do undesirable things
}

export function ok() {
  // do ok things
}
```

```js
// my-module.mjs

/* eslint no-restricted-imports: [error, {
  paths: [{
    name: external-module,
    importNames: [bad]
  }]
}] */

import { bad } from 'external-module'; // reports an error
```

An error would still be reported after making the following change:

```js
// my-module.mjs

/* eslint no-restricted-imports: [error, {
  paths: [{
    name: external-module,
    importNames: [bad]
  }]
}] */

import { bad as someAlias } from 'external-module'; // still reports an error
```

In theory, the user could "fix" the problem by modifying the external module to this:

```diff
// external-module.mjs 

export function bad() {
  // do undesirable things
}

export function ok() {
  // do ok things
}

+export const badAlias = bad;
```

```diff
// my-module.mjs

/* eslint no-restricted-imports: [error, {
  paths: [{
    name: external-module,
    importNames: [bad]
  }]
}] */


-import { bad } from 'external-module';
+import { badAlias } from 'external-module';
```

...and then the rule would no longer report an error. However, would effectively defeat the point of turning on the rule, so it would not be recommended.

Reply:
ok, so for named exports, I can see that. However, the user can "fix" it solely on the importing side by doing:
```js
import * as Module from 'external-module';
const { bad } = Module;
```

Separately, does this config option even make any sense for default imports?

Reply:
I think `import *` is disallowed whenever there are any names restricted.

Default imports can be restricted with `importNames: ["default"]`.

Reply:
@ljharb No, importing the full namespace is covered by this rule (see [this test case](https://github.com/eslint/eslint/blob/master/tests/lib/rules/no-restricted-imports.js#L247-L260)).

Reply:
Makes sense.

It seems, still, like `default` should be the only `importNames` item that should warn on the default export being imported?

Reply:
I'm confused, could you provide an example?

Reply:
`import foo from 'path'` shouldn't warn when "foo" is in `importNames`, because "foo" isn't being imported - the default is.

However, if "default" is in `importNames`, then it should be impossible to import the default from the path, under any name.

Reply:
I agree. I think that's the current behavior.

Reply:
Thanks for clarifying :-)